### PR TITLE
Use ValueStringBuilder to fix allocations in ToString() methods of XdsDateTime and XsdDuration

### LINKED ISSUE #64853: Several classes in System.Private.Xml are creating StringBuilders when formatting simple types like DateTime,TimeSpan
When I checked how much one request to our API costs found this callstack
image

This makes me curious why StringBuilder was used to format DateTime as in most cases it will be a small set of chars.
When looking at code of XmlConvert, I've found several things that can be optimized:

- ToString with TimeSpan is creating XsdDuration and calling it's ToString() which creates StringBuilder(20)
- ToString with DateTime, same with XsdDateTime
- ToString with DateTimeOffset -> XsdDateTime

I would say that whole class XmlConvert should be checked and optimized(maybe using Span<>) or ValueStringBuilder.
Please ignore if you already have plans for this or if you think that it doesn't worth.

### PR DESCRIPTION
**Pull request by TrayanZapryanov**
Fixes [Issue: 64853](https://github.com/dotnet/runtime/issues/64853)

Benchmark results :
Before

|                          Method |      Mean |    Error |   StdDev |    Median |       Min |       Max |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-------:|----------:|
|               DateTime_ToString |  79.56 ns | 1.238 ns | 0.967 ns |  79.64 ns |  77.60 ns |  81.20 ns | 0.0486 |     408 B |
|         DateTime_ToString_Local | 243.25 ns | 1.684 ns | 1.576 ns | 243.55 ns | 241.02 ns | 245.91 ns | 0.0541 |     456 B |
|   DateTime_ToString_Unspecified |  76.37 ns | 0.422 ns | 0.394 ns |  76.41 ns |  75.65 ns |  77.01 ns | 0.0487 |     408 B |
| DateTime_ToString_RoundtripKind |  78.79 ns | 0.868 ns | 0.769 ns |  78.70 ns |  77.46 ns |  80.15 ns | 0.0485 |     408 B |
|               TimeSpan_ToString |  62.96 ns | 0.485 ns | 0.454 ns |  62.88 ns |  62.09 ns |  63.75 ns | 0.0200 |     168 B |

Using SpanFormattable: 

|                          Method |      Mean |    Error |   StdDev |    Median |       Min |       Max |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-------:|----------:|
|               DateTime_ToString | 129.07 ns | 1.138 ns | 1.009 ns | 129.34 ns | 125.84 ns | 129.72 ns | 0.0131 |     112 B |
|         DateTime_ToString_Local | 291.06 ns | 4.934 ns | 4.615 ns | 288.28 ns | 287.16 ns | 300.48 ns | 0.0139 |     120 B |
|   DateTime_ToString_Unspecified | 127.20 ns | 1.123 ns | 1.051 ns | 127.49 ns | 123.74 ns | 127.97 ns | 0.0134 |     112 B |
| DateTime_ToString_RoundtripKind | 123.78 ns | 0.787 ns | 0.736 ns | 123.85 ns | 121.52 ns | 124.67 ns | 0.0132 |     112 B |
|               TimeSpan_ToString |  40.70 ns | 0.105 ns | 0.098 ns |  40.67 ns |  40.53 ns |  40.88 ns | 0.0066 |      56 B |

Latest:

|                          Method |      Mean |    Error |   StdDev |    Median |       Min |       Max |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-------:|----------:|
|               DateTime_ToString |  54.50 ns | 0.304 ns | 0.270 ns |  54.55 ns |  53.98 ns |  54.99 ns | 0.0096 |      80 B |
|         DateTime_ToString_Local | 187.15 ns | 2.430 ns | 2.029 ns | 187.60 ns | 180.93 ns | 189.17 ns | 0.0099 |      88 B |
|   DateTime_ToString_Unspecified |  49.17 ns | 0.523 ns | 0.464 ns |  49.24 ns |  47.62 ns |  49.50 ns | 0.0095 |      80 B |
| DateTime_ToString_RoundtripKind |  53.01 ns | 0.768 ns | 0.718 ns |  53.34 ns |  51.73 ns |  53.67 ns | 0.0095 |      80 B |
|               TimeSpan_ToString |  41.16 ns | 0.206 ns | 0.192 ns |  41.22 ns |  40.55 ns |  41.31 ns | 0.0065 |      56 B |

### CODE REVIEW

**Comment by danmoseley** [comment]
Could you please check that the unit tests cover these paths?

**Comment by danmoseley** [comment]
Could you change the title to something descriptive? It is more useful than a number.

**Comment by TrayanZapryanov** [comment]
> Could you please check that the unit tests cover these paths?

Looks like existing tests already found a problem :)

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs:683]:
```suggestion
        private static void IntToCharArray(Span<char> text, int start, int value, int digits)
```

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs:693]:
```suggestion
        private static void ShortToCharArray(Span<char> text, int start, int value)
```

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:706]: You can move:
https://github.com/dotnet/runtime/blob/d936a661c56f742b1660279aacb9a47d4ba1bd30/src/libraries/System.Private.CoreLib/src/System/Text/ValueStringBuilder.AppendFormat.cs#L46-L56
into:
https://github.com/dotnet/runtime/blob/d936a661c56f742b1660279aacb9a47d4ba1bd30/src/libraries/Common/src/System/Text/ValueStringBuilder.cs
and then just use that.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs:690]: Do we even need this helper anymore?  Can we instead use int.TryFormat?

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs:693]: Also these helper names don't make sense anymore.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs:517]: Can this instead just be:
```C#
vsb.AppendSpanFormattable(Year);
vsb.Append('-');
vsb.AppendSpanFormattable(Month);
```
?
Similar question for the other such blocks below.

**Comment by danmoseley** [comment]
Would you mind evaluating whether there are unit tests for the areas you're changing? This code is old, and the tests are old too so there is some possibility that it is not properly tested as a regression might slip though.

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs:706]: @stephentoub This look like a pretty big/general change and I do not prefer to do it here. Do you think if is is a problem to add partial class only in this project with the same name of the method, so if somebody is brave enough and do the proposed change - will delete my class as it will fail in compile time ?

File: src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/System/Xml/XmlConvert.cs]: Method removed

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs]: done

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs]: removed and using AppendSpanFormattable

File: src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: Thanks, but I didn't mean to separate it out into its own file.  I meant just move the function to the primary ValueStringBuilder.cs file.

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs
Comment (stephentoub) [inline src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs:652]: Nit: these braces and this extra level of indenting aren't necessary, are they?  Why add them?

File: src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: I've tried, but it was not so easy as some things stopped building.
Can we do this in another PR ?

File: src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDuration.cs
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDuration.cs:354]: Sure, I will

File: src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs
Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: C:\Work\Github\TZ_runtime\src\libraries\Common\src\System\Text\ValueStringBuilder.cs(276,110): error CS0246: The type or namespace name 'ISpanFormattable' could not be found (are you missing a using directive or an assembly reference?) [C:\Work\Github\TZ_runtime\src\libraries\System.Text.RegularExpressions\gen\System.Text.RegularExpressions.Generator.csproj]  System.Text.Json.SourceGeneration.Roslyn3.11 -> C:\Work\Github\TZ_runtime\artifacts\bin\System.Text.Json.SourceGeneration.Roslyn3.11\Debug\netstandard2.0\System.Text.Json.SourceGeneration.dll

Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: Thanks.  You should be able to fix that by adding this:
```C#
    internal interface ISpanFormattable : IFormattable
    {
        bool TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format, IFormatProvider? provider);
    }
```
here:
https://github.com/dotnet/runtime/blob/931b70a51abb3261e3a19b4240f9757473d4d397/src/libraries/System.Text.RegularExpressions/gen/Stubs.cs#L35-L37
or something along those lines.

Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: @stephentoub 
Once I've done it - release/debug build succeeds, but I say 22 projects which are linking ValueStringBuilder.
The next problem is that once I open src\libraries\System.Private.Xml\System.Private.Xml.sln, now build of tests is failing.
Again with same exception.
Should I made the trick with stubs everywhere I find linked this cs ?

Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: > The next problem is that once I open src\libraries\System.Private.Xml\System.Private.Xml.sln, now build of tests is failing.
Again with same exception.

Can you share that error?

Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: Same exception, but for project \src\libraries\Common\tests\TestUtilities\TestUtilities.csproj
9>C:\Work\Github\TZ_runtime\src\libraries\Common\src\System\Text\ValueStringBuilder.cs(276,110,276,126): error CS0246: The type or namespace name 'ISpanFormattable' could not be found (are you missing a using directive or an assembly reference?)
9>Done building project "TestUtilities.csproj" -- FAILED.

Comment (stephentoub) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: Ugh, that's annoying.  The common TestUtilities project is pulling in the product code PathInternal.Windows.cs which in turn is using ValueStringBuilder.cs.  We should probably fix that to remove that dependency, but that's not something to be done in this PR.  So, yes, let's add that internal stub into TestUtilities as well.  Does this show up anywhere else?  If it goes beyond those two places, we'll come up with something else.

Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs:20]: Something strange happens. When I build Release it is successful.
Command that I am executing is according build guide: build.cmd clr+libs -rc Release
But if I tri to build tests and to run them with : build.cmd -subset libs.tests -test -rc Release
Then I see errors, also if I open solutions in VS.
Next exception is:
1>C:\Work\Github\TZ_runtime\src\libraries\Common\src\System\Text\ValueStringBuilder.cs(276,110,276,126): error CS0246: The type or namespace name 'ISpanFormattable' could not be found (are you missing a using directive or an assembly reference?)
1>Done building project "System.Text.Encodings.Web.csproj" -- FAILED.

Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs]: You can check if this is what you have in mind.
Also other comments resolved - just the name of new file is not so good, but it is descriptive :)

Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs]: @stephentoub One more bad news :)
Using AppendSpanFormattable is slower than original implementation as we use "D4" and "D2" format. And this goes to slow formatting of int : TryFormatInt32Slow

Here are results :
SpanFormattable: 
|                          Method |      Mean |    Error |   StdDev |    Median |       Min |       Max |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-------:|----------:|
|               DateTime_ToString | 129.07 ns | 1.138 ns | 1.009 ns | 129.34 ns | 125.84 ns | 129.72 ns | 0.0131 |     112 B |
|         DateTime_ToString_Local | 291.06 ns | 4.934 ns | 4.615 ns | 288.28 ns | 287.16 ns | 300.48 ns | 0.0139 |     120 B |
|   DateTime_ToString_Unspecified | 127.20 ns | 1.123 ns | 1.051 ns | 127.49 ns | 123.74 ns | 127.97 ns | 0.0134 |     112 B |
| DateTime_ToString_RoundtripKind | 123.78 ns | 0.787 ns | 0.736 ns | 123.85 ns | 121.52 ns | 124.67 ns | 0.0132 |     112 B |
|               TimeSpan_ToString |  40.70 ns | 0.105 ns | 0.098 ns |  40.67 ns |  40.53 ns |  40.88 ns | 0.0066 |      56 B |

Original implementation + only using ValueStringBuilder
|                          Method |      Mean |    Error |   StdDev |    Median |       Min |       Max |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-------:|----------:|
|               DateTime_ToString |  52.40 ns | 0.230 ns | 0.204 ns |  52.44 ns |  52.07 ns |  52.83 ns | 0.0094 |      80 B |
|         DateTime_ToString_Local | 187.35 ns | 2.557 ns | 2.267 ns | 187.07 ns | 183.90 ns | 192.78 ns | 0.0099 |      88 B |
|   DateTime_ToString_Unspecified |  53.19 ns | 0.464 ns | 0.412 ns |  53.28 ns |  51.94 ns |  53.59 ns | 0.0094 |      80 B |
| DateTime_ToString_RoundtripKind |  54.61 ns | 0.370 ns | 0.328 ns |  54.67 ns |  53.57 ns |  54.93 ns | 0.0095 |      80 B |
|               TimeSpan_ToString |  55.19 ns | 0.300 ns | 0.266 ns |  55.23 ns |  54.49 ns |  55.57 ns | 0.0066 |      56 B |

You can use [Performance PR](https://github.com/dotnet/performance/pull/2243) for benchmark

Comment (TrayanZapryanov) [reply src/libraries/System.Private.Xml/src/Misc/ValueStringBuilder.AppendFormat.cs]: @stephentoub I've pushed changes with previous formatting for D2/4/X of integers as they are much faster. But kept changes for split of AppendSpanFormattable as they seems useful and can be used in other places. Now according benchmarks looks like we have memory + time wins. Are you okay with this changes or I should add something more ?

**Comment by TrayanZapryanov** [comment]
@danmoseley Is there something more that is expected from me ?

**Comment by danmoseley** [comment]
@stephentoub was your feedback satisfactorily addressed?

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
Thanks.

**Comment by TrayanZapryanov** [comment]
@danmoseley Once I have one approval, what's next? I see some build errors , but build was not so stable at this time. Should I somehow retrigger? Also who is merging this PR, me or somebody from the team?
Sorry for so stupid questions, but this is my first contribution.

**Comment by stephentoub** [comment]
Thanks, @TrayanZapryanov.  Not stupid questions.  We handle merging.  I re-ran the failed leg and it passed.  Thanks for the contribution.

**Pull request merged**
